common/src/event.rs

```diff
@@ -0,0 +1,203 @@
+
+use codec::{Decode, Encode};
+use gear_core::ids::MessageId;
+use primitive_types::H256;
+use scale_info::TypeInfo;
+
+/// Programs entry for messages.
+///
+/// Same as `gear_core::message::DispatchKind`,
+/// but with additional info about reply.
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo)]
+pub enum Entry {
+    /// Init entry point.
+    Init,
+    /// Handle entry point.
+    Handle,
+    /// Handle reply entry point.
+    Reply(MessageId),
+}
+
+/// Status of dispatch dequeue and execution.
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo)]
+pub enum DispatchStatus {
+    /// Dispatch was dequeued and succeed with execution.
+    Success,
+    /// Dispatch was dequeued and failed its execution.
+    Failed,
+    /// Dispatch was dequeued and wasn't executed.
+    /// Occurs if actor no longer exists.
+    NotExecuted,
+}
+
+/// Behavior of types, which represent runtime reasons for some chain actions.
+pub trait RuntimeReason: Sized {
+    /// Converter into composite reason type: not only runtime, but system also.
+    fn into_reason<S: SystemReason>(self) -> Reason<Self, S> {
+        Reason::Runtime(self)
+    }
+}
+
+// Empty implementation for `()` to skip requirements.
+impl RuntimeReason for () {}
+
+/// Behavior of types, which represent system reasons for some chain actions.
+pub trait SystemReason: Sized {
+    /// Converter into composite reason type: not only system, but runtime also.
+    fn into_reason<R: RuntimeReason>(self) -> Reason<R, Self> {
+        Reason::System(self)
+    }
+}
+
+// Empty implementation for `()` to skip requirements.
+impl SystemReason for () {}
+
+/// Composite reason type for any action happened on chain.
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo)]
+pub enum Reason<R: RuntimeReason, S: SystemReason> {
+    /// Runtime reason variant.
+    ///
+    /// This means that actor manually forced some action,
+    /// which this reason explains.
+    Runtime(R),
+    /// System reason variant.
+    ///
+    /// This means that system automatically forced some action,
+    /// which this reason explains.
+    System(S),
+}
+
+/// Runtime reason for messages waiting.
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeReason)]
+pub enum MessageWaitedRuntimeReason {
+    /// Program called `gr_wait` while executing message.
+    WaitCalled,
+}
+
+/// System reason for messages waiting.
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, SystemReason)]
+pub enum MessageWaitedSystemReason {
+    /// Program hadn't finished initialization and can not
+    /// process received message yet.
+    ProgramIsNotInitialized,
+}
+
+/// Composite reason for messages waiting.
+pub type MessageWaitedReason = Reason<MessageWaitedRuntimeReason, MessageWaitedSystemReason>;
+
+/// Runtime reason for messages waking.
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeReason)]
+pub enum MessageWakenRuntimeReason {
```

Comment:
`MessageWoken` or `MessageAwakened` maybe?

Reply:
Renamed to `MessageWoken`